odoo/tools/populate.py

```diff
@@ -0,0 +1,84 @@
+import random
+from odoo.tools import pycompat
+
+
+# format str
+# compat {** ... }
+
+def Random(seed):
+    r = random.Random()
+    r.seed(seed, version=2)
+    return r
+
+
+def format_str(val, counter, values): # todo format
+    if isinstance(val, str):
+        return val.format(counter=counter, values=values)
+    return val
+
+
+def chain_generators(fields_generates, model_name):
+    generate = root_generate()
+    for (fname, field_generate) in fields_generates:
+        generate = field_generate(generate, fname, model_name)
+    return generate
+
+
+def root_generate():
+    yield {}, False
+    while True:
+        yield {}, True
```

Comment:
What about having a pseudo-field `__complete` in the dict?  This would simplify the interface of generators: they simply generate dicts.  The pseudo-field will then be popped out of the dict by `_populate_database` before passing it to `create`.
```suggestion
def root_generate():
    yield {'__complete': False}
    while True:
        yield {'__complete': True}
```
We can actually manage a record counter that way, too.

Reply:
Good idea, I'll try this